eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/NewEpoch.hs

```diff
@@ -147,7 +147,7 @@ newEpochTransition ::
 newEpochTransition = do
   TRC
     ( _,
-      src@(NewEpochState (EpochNo eL) _ bcur es ru _pd),
+      ~src@(NewEpochState (EpochNo eL) _ bcur es ru _pd),
```

Comment:
Ah, darn, you're complete right. Good catch, and I like the suggested fix

Reply:
There are a handful of design choices here. The only thing I'm relatively confident/strongly opinionated about is that if we export a constructor (via the bidirectional pattern) that _implicitly_ sets the field to `SNothing`, then that's Chekhov's footgun.

My other suggestion is to use a `type` or `data` family for the new field along these lines.

```data NewEpochState era = ... { ... stashedAVVMAddresses :: StashedAVVMAddresses era }

type family StashedAVVMAddresses era where
  StashedAVVMAddresses (ShelleyEra c) = UTxO era
  StashedAVVMAddresses era = ()
```

Reply:
> Providing unidirctional pattern synonym without record fields (thus also not renaming original fields) and not exporting the constructor would force folks to use record update syntax.

It would, but it would also force anywhere that used the `NewEpochState` as a record pattern to have to change, since the record selectors cannot share names.

Reply:
@nfrisby I really like the close type family solution!